### DISCUSSION TITLE
datepicker mode date works wrong (TIMEZONE problem)

### DISCUSSION THREAD

**Discussion by SledgehammerPL** [wintercms/winter#786] [Q&A] [answered]
execution_date:
        label: 'valid_from'
        mode: date
        span: auto
        type: datepicker

In my database this table record is type Date.

If I set eg. 03.12.2022 it tries to set 03.12.2022 00:00:00 but GMT, but my TZ is +1 so it sets to 02.12.2022 23:00:00 (but in database writes 02.12.2022). 

I think i should somewhere set TIMEZONE to my TZ (the same as in database) but WHERE?

I found timezone in config/app.php, config/cms.php - but i want to stay with UTC.... But user can be still Warsaw/Europe. But this is stupid - date should be date (no matter of TZ)

My postgresql database was wrong set to Europe/Warsaw - but I changed in conf , altered all databases to UTC. Still the same...

I checked what is sent to postgresql:

2022-12-03 21:00:14.118 CET [1744800] xxxxxxx LOG:  execute pdo_stmt_0000000a: update "party_structure_document_items" set "document_item_execution_date" = $1 where "document_item_id" = $2
2022-12-03 21:00:14.118 CET [1744800] xxxxxxx DETAIL:  parameters: $1 = '2022-12-02', $2 = '20574'

while the parameter $1 should be 2022-12-03... So the problem is not with postgresql  -> only WinterCMS sends wrong date (zeroed)than converted to UTC. 

In modules/backend/formwidgets/DatePicker.php 

$this->getLoadValue() is already wrong!!! (date -1 day)

checked:

```
 public function prepareVars()
    {
        if ($value = $this->getLoadValue()) {
            Log::info('0 value:'.$value);  # here $value is expected date minus 1 day
            $value = DateTimeHelper::makeCarbon($value, false);

            if (!($value instanceof Carbon)) {
                throw new ApplicationException(sprintf('"%s" is not a valid date / time value.', $value));
            }

            if ($this->mode === 'date' && !$this->ignoreTimezone) {
                $backendTimeZone = \Backend\Models\Preference::get('timezone');
                Log::info('jestem  timezone '.$backendTimeZone.' value:'.$value);
                $value->setTimezone($backendTimeZone);
                Log::info('2 value:'.$value);
                $value->setTime(0, 0, 0);
                Log::info('3 value:'.$value);
                $value->setTimezone(Config::get('app.timezone'));
                Log::info('4 value:'.$value);
            }
            $value = $value->toDateTimeString();
        }
(...)

```
Here is also wrong:

modules/backend/classes/FormWidgetBase.php

public function getLoadValue()
    {
        if ($this->formField->value !== null) {
            Log::info('this->formField->value'.$this->formField->value); #here is wrong
            return $this->formField->value;
        }

        $defaultValue = !$this->model->exists
            ? $this->formField->getDefaultFromData($this->data ?: $this->model)
            : null;
        return $this->formField->getValueFromData($this->data ?: $this->model, $defaultValue);
    }

**Comment by SledgehammerPL** [comment, marked as answer]
ignoreTimezone: true

This is so stupid in mode: date that I can't even imagine who and why invent this...

**Reply by LukeTowers** [reply]
It's bitten me before and been discussed before, can't remember the outcome of the discussions. @bennothommo do you have any input on just always setting ignoreTimezone true when mode: date (or should we do the original solution proposed - I think - which was to stop the JS from sending a time value at all to the server when in mode date)

**Reply by bennothommo** [reply]
Here's the threads that I recall it being talked about:

- https://github.com/octobercms/october/issues/5072
- https://github.com/octobercms/october/pull/4111#issuecomment-628069893
- https://github.com/octobercms/october/pull/5080

I did suggest that we implicitly enable `ignoreTimezone` for the `date` or `time` modes, but I think Sam was right in that a `DATE` or `TIME` column in the database will be coerced to a Carbon instance in Laravel - not sure if it's the same with Winter. There is a lot of potential for gotchas if this is the case - since the Carbon instances, the Backend user's timezone, the app's timezone and the database timezone all could potentially play a factor in the dates or times getting skewed.
